KAFKA-7503: Connect integration test harness

wicknicks opened the pull request:
Expose a programmatic way to bring up a Kafka and Zk cluster through Java API to facilitate integration tests for framework level changes in Kafka Connect. The Kafka classes would be similar to KafkaEmbedded in streams. The new classes would reuse the kafka.server.KafkaServer classes from :core, and provide a simple interface to bring up brokers in integration tests. 

Signed-off-by: Arjun Satish <arjun@confluent.io>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

wicknicks committed:
Connect integration test harness (WIP)

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: Fix checkstyle issues

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: add util functions to produce and consume records

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: Add monitorable sink connector

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Rename variable to producer

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Add some javadocs

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
DLQ integration test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Wait for only limited time for records to come through

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Use REST endpoint to start/delete connectors

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Use distributed herder to bring up connect cluster

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Make checkstyle happy and setup and tear down clusters with @Before and @After

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Clean up ConnectIntegrationTest and reset COUNTER in stop

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch @kkonstantine @ewencp any thoughts on this PR? thanks!

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final Logger log = LoggerFactory.getLogger(ConnectIntegrationTest.class);
+
+    private final EmbeddedConnectCluster connect = new EmbeddedConnectCluster(ConnectIntegrationTest.class);
```
IIUC, we don't allow the tests to be run in parallel, because the ports are pre-defined. Is that right? If we can run in parallel, then should this be non-final and initialized in `setup()`?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        workerProps.put("bootstrap.servers", kafka().bootstrapServers());
+        workerProps.put("rest.host.name", "localhost");
+
+        putIfAbsent(workerProps, "group.id", "connect-integration-test-" + clusterName);
```
Why use strings here rather than the configuration constants? Do we think not using the constants will help with verifying backward compatibility regressions?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+
+    private static final Logger log = LoggerFactory.getLogger(EmbeddedConnectCluster.class);
+
+    private static final int NUM_BROKERS = 1;
```
Will we always want to test with only 1 embedded broker? Should this be a default that we use most of the time?

wicknicks replied:
yes, this should be a default. and should be overridden by the constructor parameters passed to this. 

would we need multiple brokers for integration tests? I did this thinking that would be more of "system tests" territory.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        putIfAbsent(workerProps, "key.converter", "org.apache.kafka.connect.json.JsonConverter");
+        putIfAbsent(workerProps, "value.converter", "org.apache.kafka.connect.json.JsonConverter");
+        putIfAbsent(workerProps, "key.converter.schemas.enable", "false");
+        putIfAbsent(workerProps, "value.converter.schemas.enable", "false");
```
If I use the Avro converter, then these will be added even though they make no sense. Perhaps we should simply check the value of the `key.converter` and `value.converter`, and only set these if we're using the corresponding JSON converter.

wicknicks replied:
good point. should we use StringConverter as defaults (if user has not passed them in `workerProps`)?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+
+        RestServer rest = new RestServer(config);
+        advertisedUrl = rest.advertisedUrl();
+        String workerId = advertisedUrl.getHost() + ":" + advertisedUrl.getPort();
```
I forget -- does Connect print the URL? If not, then it'd be nice to have a DEBUG (or INFO?) statement here reminding the developer what the URL is. WDYT?

wicknicks replied:
The rest server prints out this URL here: 

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java#L217

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        this(klass.getName());
+    }
+
+    public EmbeddedConnectCluster(String name) {
```
We don't want to go overboard, but since we're going to be using this in lots of downstream tests it'd be worth thinking about the public API. Have you thought about a builder API to make it easier to choose between different options rather than find a constructor that fits? Doing that would make it easier to add other options in the future, such as using multiple brokers and/or workers (e.g., for stopping some of them during tests), more easily setting individual properties, etc., without having to create a boatload of constructors or use setter methods that shouldn't be called after the workers are started.

```
EmbeddedConnectCluster.builder()
                      .name("MyCluster")
                      .withAvroConverters(schemaRegistryUrl)
                      .withProducerProp("batch.size", 10)
                      .withWorkers(3)
                      .withBrokers(3)
                      .build()
                      .start();
```

kkonstantine replied:
Or just accept a `Map<String, String> props`? 
That's pretty symmetric to what `EmbeddedKafkaCluster` uses with `Properties`. We just use a map instead of `Properties` in Connect.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+     * Stop the connect cluster and the embedded Kafka and
+     */
+    public void stop() {
+        after();
```
Can this be safely called more than once?

wicknicks replied:
yes, this can be called safely multiple times.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+                ConnectUtils.lookupKafkaClusterId(config), statusBackingStore, configBackingStore,
+                advertisedUrl.toString());
+        connect = new Connect(herder, rest);
+        connect.start();
```
Is there any way to subclass `DistributedWorker` to avoid duplicating the code and risk not keeping this class in sync?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+            int status = executePut(url, content);
+            if (status >= 400) {
+                throw new IOException("Could not execute PUT request. status=" + status);
+            }
```
How about writing a simple test that checks the behavior when a request to create a connector is performed with an invalid connector? Seems like it might be hard to discern the exact failure from a simple IOException.

rhauch reviewed (commented):
This is really great, @wicknicks! A few questions inline below.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,98 @@
+
+    private static final Logger log = LoggerFactory.getLogger(MonitorableSinkConnector.class);
+
+    public static final AtomicInteger COUNTER = new AtomicInteger();
```
I think `CountDownLatch` would be preferable so that we don't call `sleep` in any place in the code. 

Feel free to declare more than one if you have more than one waiting conditions.

wicknicks replied:
it's not too straightforward putting a Latch here since those can't be reset, and we would have to recreate those for every test.

kkonstantine replied:
And why not allocating a new one in every test?
I really believe that `sleep` eventually leads to flaky tests and also introduces unneeded, constant delays, that eventually add up and make tests unbearable to sit and wait for to finish.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public Class<? extends Task> taskClass() {
+        return SimpleTestSinkTask.class;
```
hmm, why not `MonitorableSinkTask`? It's a common convention.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,98 @@
+        return new ConfigDef();
+    }
+
+    public static class SimpleTestSinkTask extends SinkTask {
```
ask above too. `MonitorableSinkTask`?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,98 @@
+
+        @Override
+        public String version() {
+            return null;
```
How about: `return "unknown";`?

kkonstantine replied:
It's a common default in connectors

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+    protected void before() throws IOException {
+        kafkaCluster.before();
+        startConnect();
+        log.info("Started connect at {} with kafka cluster at {}", restUrl(), kafka().bootstrapServers());
```
Kafka, ZooKeeper use upper case already. 
`Connect` will be easier not to confuse with the verb `connect` even if the context helps

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    public void startConnect() {
+        log.info("Starting standalone connect cluster..");
```
`standalone` is an overloaded term already. Not sure if it was referring to an implementation based on standalone herder vs distributed that is now used. But it seems unnecessary here.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    public void startConnector(String connName, Map<String, String> connConfig) throws IOException {
+        String url = String.format("http://localhost:8083/connectors/%s/config", connName);
```
I'd declare host and port somewhere, even though they are not meant to be configurable yet (basically the port)

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        try {
+            String content = mapper.writeValueAsString(connConfig);
+            int status = executePut(url, content);
+            if (status >= 400) {
```
There should be an enum for `400`

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    private void putIfAbsent(Map<String, String> props, String propertyKey, String propertyValue) {
```
seems like `static` too, right?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,226 @@
+        httpCon.setDoOutput(true);
+        httpCon.setRequestProperty("Content-Type", "application/json");
+        httpCon.setRequestMethod("PUT");
+        OutputStreamWriter out = new OutputStreamWriter(httpCon.getOutputStream());
```
This can also use try-with-resources and avoid calling `close` explicitly

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,291 @@
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+
+public class EmbeddedKafkaCluster extends ExternalResource {
```
If we are copying as-is from streams, maybe we should add a comment to state that.

kkonstantine replied:
Also, if that's the case, I'll skip minor comments for the same of uniformity.

kkonstantine commented on connect/runtime/src/test/resources/log4j.properties:
```diff
@@ -14,10 +14,20 @@
 ##
-log4j.rootLogger=OFF, stdout
+# Root logger option
+log4j.rootLogger=INFO, stdout
```
How about at least `DEBUG`?

kkonstantine commented on connect/runtime/src/test/resources/log4j.properties:
```diff
@@ -14,10 +14,20 @@
 
+log4j.logger.org.reflections=ERROR
 log4j.logger.org.apache.kafka=ERROR
+log4j.logger.org.apache.kafka.connect=INFO
```
how about at least `DEBUG`. Also why override if it matches the `rootLogger`'s level?

wicknicks replied:
this was an accidental commit. reverting it. keeping INFO make running `./gradlew test` very noisy. logging is turned of by default.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+
+        // consume all records from test topic
+        log.info("Consuming records from test topic");
+        for (ConsumerRecord<byte[], byte[]> recs : connect.kafka().consumeNRecords(20, 5000, "test-topic")) {
```
factoring out constants as variables will help with quick understanding of the values passed, as well as with replication of the test elsewhere.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+        connect.kafka().createTopic("test-topic");
+
+        // produce some strings into test topic
+        for (int i = 0; i < 10; i++) {
```
should also `10` be a self-explaining variable?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+            log.debug("Consumed record ({}, {}) from topic {}", new String(recs.key()), new String(recs.value()), recs.topic());
+        }
+
+        Map<String, String> confs = new HashMap<>();
```
I often see `configs` and `props`. In case you want to borrow naming from elsewhere.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+        connect.startConnector("simple-conn", confs);
+
+        int attempts = 0;
+        while (attempts++ < 5 && MonitorableSinkConnector.COUNTER.get() < 18) {
```
The numbers alone are hardly self explaining when looking quickly. Especially `18`
See also comment about using a `CountDownLatch` and avoid constants and `sleep`

kkonstantine replied:
you can time wait on the latch if you want to log progress

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+
+        // consume failed records from dead letter queue topic
+        log.info("Consuming records from test topic");
+        for (ConsumerRecord<byte[], byte[]> recs : connect.kafka().consumeNRecords(2, 5000, DLQ_TOPIC)) {
```
I think N in `consumeNRecords` is redundant. The fact that it accepts an argument (that ideally has a name vs a value) should be enough. `consume`?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+        // consume failed records from dead letter queue topic
+        log.info("Consuming records from test topic");
+        for (ConsumerRecord<byte[], byte[]> recs : connect.kafka().consumeNRecords(2, 5000, DLQ_TOPIC)) {
+            log.info("Consumed record ({}, {}) from dead letter queue topic {}", new String(recs.key()), new String(recs.value()), DLQ_TOPIC);
```
Maybe it's not obvious that inside the parenthesis you'll have key and value.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+        @Override
+        public R apply(R record) {
+            long val = Long.parseLong(String.valueOf(record.value()));
+            if (val == 7) {
```
Not obvious what this condition symbolizes with just the value `7`. Use a variable name?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,145 @@
+
+        @Override
+        public void configure(Map<String, ?> configs) {
+            log.info("Configuring {}.", this.getClass());
```
Do you prefer the simple class name here? Not sure.

kkonstantine reviewed (commented):
I really anticipate using this! Thanks @wicknicks !
Left a round of comments.

wicknicks committed:
Address reviewer comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Reset log4j.properties to original state

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Clean up imports

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch @kkonstantine this is ready for a second pass. thanks!

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,94 @@
+    private static final Logger log = LoggerFactory.getLogger(ConnectIntegrationTest.class);
+
+    private static final int NUM_RECORDS_PRODUCED = 2000;
+    private static final int CONSUME_MAX_DURATION_MILLIS = 5000;
```
nit: Everywhere I remember we use `_MS` suffix for time unit. You might not want to introduce yet another way (in case you grep or search some time) `CONSUME_MAX_DURATION_MS`?

kkonstantine reviewed (commented):
Thanks for addressing reviews @wicknicks 
One nit and one question still from me.

wicknicks committed:
Align variable name to convention

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Use a latch to wait for connector to sink records

Signed-off-by: Arjun Satish <arjun@confluent.io>

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java:
```diff
@@ -53,62 +53,25 @@
 
-    public static void main(String[] args) throws Exception {
+    private Time time = Time.SYSTEM;
+    private long initStart = time.hiResClockMs();
```
should this be `final` since it's already global? Are you planning to reassign it? (probably not given its name)

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,99 @@
+        connect.kafka().createTopic("test-topic");
+
+        // produce some strings into test topic
+        for (int i = 0; i < NUM_RECORDS_PRODUCED / 2; i++) {
```
Just be ready for odd number of `NUM_RECORDS_PRODUCED` this not to do what you'd think it'd do (will add one more record I guess)

I wonder why it's important to have different prefix (`hello-` and `world-`) vs a common one such as `hello-world-`. I see this is the value not the key.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,99 @@
+        connect.startConnector("simple-conn", props);
+
+        MonitorableSinkConnector.MonitorableSinkTask task = MonitorableSinkConnector.TASKS.get("simple-conn-0");
+        for (int i = 0; i < 10 && task == null; i++) {
```
What is `10` here? Retries? Let's give it a name.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,126 @@
+
+        public void awaitRecords(int consumeMaxDurationMs) throws InterruptedException {
+            if (latch == null) {
+                throw new IllegalStateException("latch was not initialized");
```
The message is true here. But also, not clear, if you don't look at the implementation of this class. I feel something like "Task has not been started" might make more sense and not expose impl details. 

nit: (there's no no-upper-case policy from what I know in our messages :) )

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,126 @@
+                throw new IllegalStateException("latch was not initialized");
+            } else {
+                if (!latch.await(consumeMaxDurationMs, TimeUnit.MILLISECONDS)) {
+                    throw new DataException("Insufficient records seen by task");
```
If I'm not mistaken you know how many records you expect. 
Should we print that to the message ("Fewer than X records were consumed by the task in Y ms") ?

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,99 @@
+        MonitorableSinkConnector.MonitorableSinkTask task = MonitorableSinkConnector.TASKS.get("simple-conn-0");
+        for (int i = 0; i < 10 && task == null; i++) {
+            log.debug("Sleeping for 200 before looking up task");
+            Thread.sleep(200);
```
Here's an idea. How about locking the whole map and wait on the map itself (timed wait)? Once a thread adds a task to the map it should `notify` this map. Then you can check on the size of the map and determine when to make progress. 

Given that, you might want to use `Collection.synchronizedMap` even it doesn't expose it's internal `mutex` so you can use it. Still, it's clear you are not depending on the lock free concurrent map version but you lock the whole map. No big deal for this use case. It's not critical contention wise.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,150 @@
+        connect.kafka().createTopic("test-topic");
+
+        // produce some strings into test topic
+        for (int i = 0; i < NUM_RECORDS_PRODUCED / 2; i++) {
```
Comments on the `ConnectIntegrationTest` apply here too. Please make sure the approach in both classes is the same.

kkonstantine reviewed (commented):
Thanks for addressing my latest comments @wicknicks 
Took a fresh look and added a few more which I hope you'll find useful. I think we are close.

wicknicks committed:
Test can wait till task is created instead of busy loop

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Address other review comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Make variables final

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Remove unused imports

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@kkonstantine thanks so much for your multiple passes. I believe I've addressed all your comments (tried something a bit with the latches though). Please take a look at the changes, and let me know if they work out.

kkonstantine reviewed (approved):
I think it looks great and is a significant step forward in Connect's testing capabilities. 
Thanks for addressing all the comments in detail. 

LGTM by me.

wicknicks commented:
Thanks so much, @kkonstantine. Highly appreciate the review comments.

@rhauch I believe I have addressed all your comments as well. would you mind taking another pass at this?

wicknicks commented:
~one quick question to everyone. do you think it makes sense to move the EmbeddedConnectCluster and EmbeddedKafkaCluster to `/connect/runtime/src/main/java/org/apache/kafka/connect/tools/`? this well help us package these classes in case we want to use them to test connector code in the future. thanks!~

edit: ah, looks like the test jars to pushed to maven central as well. please ignore this comment.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/cli/ConnectDistributed.java:
```diff
@@ -118,4 +81,49 @@ public static void main(String[] args) throws Exception {
+        log.info("Scanning for plugin classes. This might take a moment ...");
+        Plugins plugins = new Plugins(workerProps);
+        plugins.compareAndSwapWithDelegatingLoader();
+        DistributedConfig config = new DistributedConfig(workerProps);
```
It might be even more valuable to break this method into multiple protected methods and to have more of the components created in this method be members. However, that's probably work for another ticket.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Map;
+
+@Category(IntegrationTest.class)
+public class ConnectIntegrationTest {
```
I think we should have JavaDoc that explains the purpose and scope of each of these IntegrationTest classes. For example, what is the scope of `ConnectIntegrationTest` and `DeadLetterQueueIntegrationTest`? When should we create additional classes versus adding methods to one of these two? If each had a brief summary of the intended scope, it becomes easier to know whether to add a method or add a new integration test class.

Or, perhaps instead this class should be renamed to more precisely define the scope.

ewencp replied:
The added javadoc helps, but I'd also just rename this. It's so generic that it can easily become a dumping ground.

More generally re: integration tests, striking the right balance can be tough -- this name is probably to broad and open ended, the dead letter queue test is probably too narrow. This integration test is validating basic happy path behavior, so I'd indicate that in the name somehow.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,169 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MonitorableSinkConnector extends TestSinkConnector {
```
JavaDoc would be nice.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    // [2018-09-25 11:23:34,562] DEBUG Waiting on task simple-conn-0 (org.apache.kafka.connect.integration.MonitorableSinkConnector:68)
+    // [2018-09-25 11:23:36,058] DEBUG Found task! (org.apache.kafka.connect.integration.MonitorableSinkConnector:70)
```
What do these comments mean?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,169 @@
+    public static class Handle {
+        private static final int MAX_WAIT_FOR_TASK_DURATION_MS = 60_000;
+
+        private final String taskName;
```
Should this be `taskId` to be more consistent with the parameters of the methods below?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,169 @@
+
+        private final String taskName;
+        private final AtomicReference<MonitorableSinkTask> taskRef = new AtomicReference<>();
+        private final CountDownLatch taskAvailable = new CountDownLatch(1);
```
How about a CountDownLatch when then the tasks have all been stopped? I know that the Handle is currently removed when the connector is stopped, but before it is stopped i might want to get the latch to make it easier to do something when a connector is actually stopped. Then, even if the connector is stopped after that, I still have the latch and can still perform my logic.

wicknicks replied:
this is addressed now. great catch!

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,298 @@
+        throw new RuntimeException("Could not find enough records. found " + consumedRecords + ", expected " + n);
+    }
+
+    public KafkaConsumer<byte[], byte[]> createConsumer(String... topics) {
```
It might be better to have a `createConsumer(Map<String, Object> props)` method and `createConsumer()` method that does everything here except the subscribe, and to leave the subscribe call (which is one method invocation) to the caller. (Probably should mention the latter in the JavaDoc for this method.)

This additional control will make it easier for users to create a consumer that fits their needs. And since this method is currently only called from within the `consume(...)` method, there's little downside in having the caller do the subscribe.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,298 @@
+        Map<String, Object> props = new HashMap<>();
+
+        props.put(GROUP_ID_CONFIG, UUID.randomUUID().toString());
+        props.put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers());
```
This should always override what the user specifies, but all of the others might be overridden by the supplied properties.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,298 @@
+        producer = new KafkaProducer<>(producerProps);
+    }
+
+    private void stop() {
```
Why is this private and not public, if `start()` is public?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,195 @@
+        return kafkaCluster;
+    }
+
+    public int executePut(String url, String body) throws IOException {
```
How about one for POST?

wicknicks replied:
do we need one right now? most of the connect API uses GET or PUT. POST is only for restarting connectors or tasks.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,195 @@
+        connect = new ConnectDistributed().startConnect(workerProps);
+    }
+
+    public void startConnector(String connName, Map<String, String> connConfig) throws IOException {
```
Might be nice to have an `updateConnector` method, since I'd probably want to have tests that check the behavior (at least within AK) to verify that connectors can be updated and they (re)start correctly.

wicknicks replied:
I think the REST endpoint for updating the connector is the same as the one for creating. we simply reuse the name. That's why I didn't add a new method for `updateConnector`.

rhauch replied:
Yeah, what if I want to test updating a connector configuration to see if my connector responds correctly? Seems valuable enough, and if it's not in the integration test framework connector projects simply can't do it. (They could implement their own, but why not have it here?)

ewencp replied:
i think the follow up here given the operation is the same could be to just rename this as `configureConnector` or something like that, such that it is not specific to whether it's the first time you are doing it (which also creates the connector) or later (which just updates)

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,195 @@
+    private final Map<String, String> workerProps;
+    private final String clusterName;
+
+    private Connect connect;
```
Just one worker? Why not allow multiple workers? We absolutely need integration tests that verify that the followers properly forward requests to the leader. Yes, you'll have to manage port numbers and maybe have a method that allows the caller to use a particular worker (the bulk of the methods in this class can just use the current worker).

wicknicks replied:
The workers will conflict over the REST port. Right now, until we can make the REST port configurable, we will need to limit this to one worker per concurrent test.

rhauch replied:
Okay, that's a good start.

ewencp replied:
not sure i understand the comment re: REST port. that is configurable already. seems like if there is a constraint re: ports it would be whether we can go back and extract the bound port since for tests you want to configure it on port 0 to avoid conflicts but then need to look up actual bound port.

we should not commit these integration tests without fixing that. not using dynamic ports has historically been a huge source of flaky tests. (i'm particularly sensitive about this issue since i spent a bunch of time fixing https://issues.apache.org/jira/browse/KAFKA-1501 way back in the day :) )

wicknicks replied:
sorry, I meant to say that Connect simply uses the port number given to it. Unlike `EmbeddedKafkaCluster`, where setting a port value of 0 will make it select any free port available on the system. I thought making that change would need a KIP. Let me know if this is not the case.

P.S: I couldn't open the link. The website is probably down atm.

ewencp replied:
not sure i understand the issue, why wouldn't the tests just override the port setting to 0 and then lookup the bound port, just like the EmbeddedKafkaCluster? Kafka also normally has a default port of 9092 in its config, but the tests override to 0, and EmbeddedKafkaCluster calls that "default" because for tests it is the only way to make it reliable and not make tests take many hours to run. This will, however, likely requiring adding a bit of code in the Jetty/Jersey REST API code to extract the bound port.

We basically can't merge this without fixing this since it breaks the common developer path of `./gradlew integrationTest` (or anything that covers all tests), which is documented as a normal workflow in README.md.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,169 @@
+
+        @Override
+        public void stop() {
+            log.info("Removing {}", HANDLES.remove(taskId));
```
Woah, I almost missed this important call to `remove(taskId)` since it's within a log message. Not sure this is a good idea.

This won't work if the # of tasks is > 1, since the first task to be stopped will remove the Handle for the connector.

wicknicks replied:
Wouldn't `taskId` be different for each task?

rhauch reviewed (commented):
Fantastic, @wicknicks! I have a few suggestions below, but all are pretty minor.

wicknicks committed:
Add some javadoc and count down on latch when task exits

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Update ConnectIntegrationTest to use multiple topic partitions and tasks

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch reviewed (approved):
@wicknicks, really nice work. This will be so valuable, and I can't wait to use it already!

wicknicks commented:
@ewencp any thoughts on this PR?

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTest.java:
```diff
@@ -0,0 +1,97 @@
+     * The test will produce and consume records, and start up a sink connector which will consume these records.
+     */
+    @Test
+    public void testProduceConsumeConnector() throws Exception {
```
I like this as a baseline integration test.

We should consider what set of happy path dimensions we want to validate here and if test parameterization or extension here would help with testing coverage.

In particular, I think there's a number of implicit decisions here: which converter is being tested, there are no transformations, that the default integration test uses distributed mode (and therefore specific offsetwriter, config source, etc), we're using connector overridden converter settings, which data types are used in the records, etc. There's always going to be a fair bit that is explicit, but some of this may be worth deciding whether to specifically test here, and to get coverage in a smaller number of tests.

In particular, things like covering many/all data types and including testing of transformation functionality are things I think we should probably try to include in a single test. Our experience elsewhere with these heavyweight "integration" tests which actually also pull up external services rather than mocking them is that they get very expensive, very fast, and that's part of why I avoided adding them for Connect initially. I think if we're doing these expensive "local system test" version of integration testing, we should try to write fewer more comprehensive tests.

As another example, this test could also validate various REST API functionality once the connector is submitted. This covers a bit of it with the create/delete, but we could also validate, e.g., the status APIs.

wicknicks replied:
@ewencp yes, agree that these tests should not be expensive. But when you say they must be comprehensive, does that mean creating multiple tests with different converters and transformation. I don't think that should be the purpose here. We need to ensure that the interfaces of these different components work well together. Plugging in different implementations of converters ensures their correctness, but that can be tested with unit tests. Here we want to make sure that all the components (through their interfaces, work correctly with each other). For example, a test could be extended to add a transformation and/or a dead letter queue to check if all of these pieces work well together (which is what `DeadLetterQueueIntegrationTest` does, btw). Variations of this test would be running the same test (move data through a connector pipeline) in the absence of the optional components and asserting the correct behavior. 

in the light of the above comment, what about calling this test as `EndToEndSinkConnectorIntegrationTest`, and moving the test from `DeadLetterQueueIntegration` into this class as a `testWithDeadLetterQueue` test method?

@rhauch @ewencp let me know what you think?

ewencp replied:
By comprehensive, I actually just meant that the integration test should evaluate things more end-to-end, i.e. that the integration of multiple components are working.

I agree that you are frequently testing interfaces, but also lots of frequently unspoken assumptions (or not enforced), e.g. even simple things like if `null` is a valid value for a particular argument or not, and whether other components calling the method will deliver them.

You can't cover the entire space of combinations, but I called out naming because I think the name says a lot about what you think you're testing. `DeadLetterQueueIntegrationTest` is named to be quite specifically targeted -- it's testing a single config/policy/*mostly* one class (`DeadLetterQueueReporter`).

The suggestion re: individual test methods is interesting as well -- this was also what I was getting at wrt this style of test being quite heavyweight and therefore better to keep broader. If it was an integration test with fast mocks (e.g. even if full Connect worker, mocked out ZK + broker), I'd be more inclined to have tests like `ErrorHandlingIntegrationTest` with separate tests for each. But given how expensive they are, why *not* just have a single test method in an `ErrorHandlingIntegrationTest` validate all the error handling cases end-to-end?

I guess high level what I'm getting at is that there's a tendency to write integration and system tests that validate individual features that are frequently effectively a single class + a couple of surrounding ones, but pull in a ton of other test surface area. I think getting the balance is especially hard with integration tests, but with system tests I've noticed some similar tests as well. But if you go that route, you end up with basically as many integration tests as you have unit tests, which shouldn't really be the case (the thing the integration test is testing should be larger than the unit test)?

wrt "testing the same thing", I think an end-to-end source connector and sink connector test where we swap out components in a parameterization that reasonably covers the space *is* a good set of integration tests (since it validates the source and sink data pipelines, integrated together, under at least a few different variants, e.g. different types of converter, transforms, etc). Some of those will ultimately land with the individual connectors as well (e.g. is the connector & upstream converters/transforms actually obeying the assumptions of interfaces that may not be explicit yet).

Another way to view this is that the javadoc should say *what is under test* and maybe even *why* we need this in addition to unit tests; currently this test's javadoc just says *what it will do*.

wicknicks replied:
yes, that makes sense. By interfaces I meant the glue code which sits between modules not necessarily just the "Java interfaces". I've made the changes based on this discussion. Ideally we could have removed the `ConnectIntegrationTest` class, but it's now renamed to `ExampleConnectIntegrationTest` with the javadoc specifying that this class is to be simply used to show how to use the integration test framework. In future, when we make enhancements, we can update this test accordingly. But otherwise we should not be adding any tests here. The `DeadLetterQueueTest` has been moved to `ErrorHandlingTest` and occurs all the error handling policies. Also, this is perfect since all the integration points of those policies are independent of each other, and we can test them in one test.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/DeadLetterQueueIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+ * the connector pipeline to fail. The bad records must be found in the dead letter queue.
+ */
+@Category(IntegrationTest.class)
+public class DeadLetterQueueIntegrationTest {
```
re: naming and how broad/narrow tests are, it might make sense to have a single test run separate connectors for each of the error handling strategies and validate them. i realize it makes the test a bit more complex and potentially a bit harder to debug, but I think for the cost of this style of integration tests, that's the right scope for a single test. (this is something I increasingly think we need to consider when writing system tests as well...)

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,181 @@
+    public static final String EXPECTED_RECORDS = "expected_records";
+    private static final Logger log = LoggerFactory.getLogger(MonitorableSinkConnector.class);
+
+    private static final Map<String, Handle> HANDLES = new ConcurrentHashMap<>();
```
if this is going to be static, should we have code somewhere to ensure we're in a good, clean state before we start a test? otherwise one bad test could leave state in here that makes all subsequent tests using it that happen to use identical task IDs see stale data. (using a `static` here seems like a reasonable tradeoff given other options for getting at the actual connector instance, but if we're forced to use a `static`, we should also be defensive)

wicknicks replied:
this is fixed. thanks! great catch!

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    public void start() throws IOException {
+        before();
```
why both names for these? I see the before/after is for `ExternalResource`. is using that interface gaining us something? i see we use it in streams too, but I'm unclear on why. i would have thought it would help us by automatically cleaning up or something, but it looks like we're still calling these manually in tests.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,199 @@
+
+    public EmbeddedConnectCluster(String name, Map<String, String> workerProps, int numBrokers, Properties brokerProps) {
+        this.workerProps = workerProps;
+        this.clusterName = name;
```
maybe we should generate these randomly rather than requiring the test to feed them in? that would guarantee uniqueness and ensure we don't have any accidental (and confusing) potential conflicts between tests.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,199 @@
+
+        putIfAbsent(workerProps, GROUP_ID_CONFIG, "connect-integration-test-" + clusterName);
+        putIfAbsent(workerProps, OFFSET_STORAGE_TOPIC_CONFIG, "connect-offset-topic-" + clusterName);
+        putIfAbsent(workerProps, OFFSET_STORAGE_REPLICATION_FACTOR_CONFIG, "1");
```
this should probably be set via some variable if we're going to extend this to support multi worker cases.

wicknicks replied:
let's make this change when we start supporting multi-worker tests.

wicknicks replied:
@ewencp, did you mean multi-worker or multi-broker?

wicknicks replied:
I don't think I understand the implication of replication factor on multi worker tests. thanks!

rhauch replied:
I think this has to do with running more than one embedded brokers, at which point you want a replication factor that is > 1.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    public String bootstrapServers() {
+        return address(brokers[0]);
```
we should use all available broker addresses here. this would be important in any test cases that might cover failure of brokers.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,314 @@
+     * @param topicConfig Additional topic-level configuration settings.
+     */
+    public void createTopic(String topic, int partitions, int replication, Map<String, String> topicConfig) {
+        replication = Math.min(replication, brokers.length);
```
we should just throw an exception if there aren't enough brokers. i think the behavior if we do this will just be confusing (something we've seen in practice too with the consumer offsets topic when there aren't enough active brokers and it gets created with too low of a replication factor).

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,314 @@
+    public ConsumerRecords<byte[], byte[]> consume(int n, long maxDuration, String... topics) {
+        long num = 1;
+        long duration = maxDuration;
+        // poll in intervals of 250millis.
```
why this breakdown of the max duration? it doesn't look like there is anything we do periodically (e.g. logging) that requires breaking this down

wicknicks replied:
I suppose this was just my way of making sure we wait for at most maxDuration *and* consume n records. I think there are some instances where we wait for maxDuration, and if consumer returns sooner with fewer records, the logic will recalculate how much time is left, and then attempt again. having said that, looks like my logic would break if n records are spread over more than `num` intervals.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,314 @@
+        return consumer;
+    }
+
+    public KafkaConsumer<byte[], byte[]> createConsumer() {
```
is this used, and if so, how? doesn't seem useful to be creating consumers that don't subscribe to anything

wicknicks replied:
this is used right below, line 312:

```
    public KafkaConsumer<byte[], byte[]> createConsumerAndSubscribeTo(String... topics) {
        KafkaConsumer<byte[], byte[]> consumer = createConsumer();
        consumer.subscribe(Arrays.asList(topics));
       ....
```

ewencp replied:
right, my point was usage outside of this class. We have multiple public createConsumer* methods here, but the only one that's actually useful externally is createConsumerAndSubscribe. Being able to also override consumer props seems potentially useful, but that should just be two methods -- one that takes only topics and another that takes topics and consumer props overrides.

wicknicks committed:
Changes to startConnector method

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Remove before() and after() methods from EmbeddedConnectCluster

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Generate clusterName randomly instead of using argument

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
bootstrapServers() returns all brokers

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Throw exception if desired replicator is unachievable

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Refactor logic for consuming up to n records in maxDuration

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Remove unused methods

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Rename tests and broaden the scope of error handling test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
Clean up handles for tasks before starting the test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp I think I've addressed all your comments. Could you please take another pass? thanks so much!

wicknicks committed:
Correct some typos

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented:
@ewencp any chance you could squeeze in a review of the latest changes?

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ErrorHandlingIntegrationTest.java:
```diff
@@ -0,0 +1,208 @@
+            assertTrue(recs.headers().toArray().length > 0);
+            assertValue("test-topic", recs.headers(), ERROR_HEADER_ORIG_TOPIC);
+            assertValue(RetriableException.class.getName(), recs.headers(), ERROR_HEADER_EXCEPTION);
+            assertValue("Error when when value='value-7'", recs.headers(), ERROR_HEADER_EXCEPTION_MESSAGE);
```
nit: `when when`. not a big deal though since this is just in the test

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,190 @@
+        public String toString() {
+            return "Handle{" +
+                    "taskId='" + taskId + '\'' +
+                    ", taskAvailable=" + taskAvailable.getCount() +
```
nit: you might want to make this a boolean based on the count. since it's a countdown latch, the count could be a bit confusing when you see this in a log or debugger

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    public static Handle taskInstances(String taskId) {
```
this being plural is a bit confusing. also, seems like you're always going to call `taskInstances(taskId).task()`. Is there a use case for not just calling `task` in this method and returning `MonitorableSinkTask`?

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,190 @@
+
+    public static void cleanHandle(String taskId) {
+        HANDLES.computeIfPresent(taskId, (k, handle) -> {
+            handle.taskAvailable.countDown();
```
this doesn't seem right. elsewhere, we're using the countdown latch to make sure that the task is actually available. when a task is cleaned up, `taskAvailable` should have already had `countDown` once, so doing it again won't do anything, right? shouldn't we just be removing the handle from `HANDLES` entirely? or doing something to reset `taskAvailable` to have a positive count?

wicknicks replied:
@ewencp: we are returning null in the lambda here to remove this handle from the map. 

countDown() is called on the latch only for the case where `cleanHandle()` is called when some thread is still waiting on a task (using `CountDownLatch#await()`). although in this class they would only wait for a fixed duration, [line 75](https://github.com/apache/kafka/pull/5516/files#diff-1ba0a6e4409701d883e6a2624e622e33R75)). 

I've added some documentation that make this clearer. thanks!

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSinkConnector.java:
```diff
@@ -0,0 +1,190 @@
+        for (int i = 0; i < maxTasks; i++) {
+            Map<String, String> config = new HashMap<>();
+            config.put("task.id", connectorName + "-" + i);
+            config.put("required", "dummy-val");
```
this doesn't seem to do anything?

ewencp reviewed (commented):
getting close. main thing left is the REST port, as it stands the standard developer CLI workflow won't work -- it consistently throws exceptions because of port conflicts. for integration tests in particular it's critical to be able to run in parallel because they are so much slower than unit tests.

wicknicks committed:
KAFKA-7503: Fix typos and remove unnecessary task props

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Rename taskInstances method

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Add documentation to clarify taskAvailable's contracts

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Let Jetty pick any available port for integration tests

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Multi worker connect cluster in integration tests

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Add counters in connector handle

This is useful when the exact expectations cannot be set for tasks
(for the purposes of integration tests, tasks can be a bit unreliable
since a rebalance can delay when a task comes up and gets ready to
consume records).

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp looks like I misinterpreted the changes needed to set the connect port dynamically. I've added commits which configure connect to pick any available (by setting `rest.port` to 0). also, we should now be able to bring up multiple connect workers. finally, the classes were refactored for readability. specifically: 
* moved out the `HANDLES` map in `MonitorableSinkConnector` into a separate class(es) to make the overall code to monitor record arrival more readable; 
*  the `MonitorableSinkConnector` connector now no longer manages latches, this is done in the handles. the tasks only call `TaskHandle` methods to count down latches on message arrival.
* the `ConnectorHandle` class records the messages arrived in all of its tasks. this is useful when we cannot rely on an exact number of messages coming into tasks (for example: in the multi worker case, the tasks can get rebalanced multiple times, and there is a race condition on which task consumes how many records).

wicknicks committed:
KAFKA-7503: Remove unnecessary task handle cleanup requests

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
retest this please

wicknicks committed:
MINOR: Change example worker props

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: Use better error message

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
TEMP: Modify log4j.properties to debug failing test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: Find free ports using ServerSocket API

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp: we have a green build (JDK 8) now. Could you please take a look at this PR, and if it looks good, can we merge it? thanks!

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorHandle.java:
```diff
@@ -0,0 +1,105 @@
+    private static final Logger log = LoggerFactory.getLogger(ConnectorHandle.class);
+    private final String connectorName;
+
+    private Map<String, TaskHandle> taskHandles = new ConcurrentHashMap<>();
```
you can `final` this

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ErrorHandlingIntegrationTest.java:
```diff
@@ -0,0 +1,219 @@
+
+        // consume all records from test topic
+        log.info("Consuming records from test topic");
+        for (ConsumerRecord<byte[], byte[]> recs : connect.kafka().consume(NUM_RECORDS_PRODUCED, CONSUME_MAX_DURATION_MS, "test-topic")) {
```
if we want to validate this, since we're not guaranteed the exact number of records from `consume`, we should probably assert the number of records consumed here. in fact, we have to take care in doing this re: retry settings in the producer, and in subsequent validation of the connector. in practice, consuming here probably isn't all that useful -- we should make strong assumptions about availability when we expect it in integration tests and just fail the `produce` phase if it doesn't meet our expectations.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:
```diff
@@ -0,0 +1,121 @@
+     * Record a message arrival at the task.
+     */
+    public void record() {
+        if (recordsRemainingLatch != null) {
```
if this isn't the case, shouldn't this be an error, similar to the case in `awaitRecords`? it may not be handled quite the same, but it should be recorded as an error somewhere so it can be propagated to the main test.

wicknicks replied:
this was the expected behavior: in some tests we only care about the total number of records seen by the connector, but in others we want a specific number to be handled by each task. The assertion is made later (if needed) by creating the latch as shown here: https://github.com/apache/kafka/pull/5516/files/#diff-04a4db29ce052d031b662738cdda1677R127

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExampleConnectIntegrationTest.java:
```diff
@@ -0,0 +1,132 @@
+        connect.configureConnector(CONNECTOR_NAME, props);
+
+        // expect all records to be consumed by the connector
+        connectorHandle.expectedRecords(NUM_RECORDS_PRODUCED);
```
Is there an ordering race condition here? Shouldn't we always do the `expectedRecords` *before* configuring the connector? Otherwise the request in `configureConnector` could complete, Connect could start the connector and tasks, and some records could be processed before the number expected & latch are reset.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExampleConnectIntegrationTest.java:
```diff
@@ -0,0 +1,132 @@
+        connectorHandle.awaitRecords(CONSUME_MAX_DURATION_MS);
+
+        // at least one task should have initalized and executed
+        assertTrue("Incorrect task count in connector", connect.getConnectorStatus(CONNECTOR_NAME).tasks().size() > 0);
```
we should know in this test exactly how many tasks there are, this should really be an `assertEquals` instead of assertTrue`. we should be as precise with these assertions as we can be.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:
```diff
@@ -0,0 +1,121 @@
+     * @param numPartitions number of partitions
+     */
+    public void partitionsAssigned(int numPartitions) {
+        for (int i = 0; i < numPartitions; i++) {
```
given the countdown latch only currently supports a single partition, we should probably trigger an error if more than that number is provided

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,304 @@
+            try {
+                worker.stop();
+            } catch (Exception e) {
+                log.error("Could not stop connect", e);
```
we shouldn't just be swallowing these, it would cause broken worker stop code to pass tests. we can either collect them and throw a wrapper or just let this propagate.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,304 @@
+        try {
+            kafkaCluster.after();
+        } catch (Exception e) {
+            log.error("Could not stop kafka", e);
```
same here re: swallowing instead of allowing to propagate.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,304 @@
+        }
+        log.debug("Found available ports: {}", Arrays.toString(ports));
+        for (int i = 0; i < connectCluster.length; i++) {
+            workerProps.put(REST_PORT_CONFIG, String.valueOf(ports[i])); // set this to zero so we pick a random port
```
comment doesn't seem to match implementation?

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,304 @@
+        }
+    }
+
+    public String endpointForResource(String resource) {
```
nit: doesn't seem like this needs to be public

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java:
```diff
@@ -0,0 +1,304 @@
+    }
+
+
+    private int findFreePort() {
```
this approach is known to cause a lot of transient failures which are difficult to track down, and they happen a lot more frequently on bogged down/heavily utilized test machines like jenkins. a lot of time and effort was spent cleaning up the previous code that tried to use this approach and which caused a ton of instability: https://issues.apache.org/jira/browse/KAFKA-1501 The proper way to implement this is allow the port 0 config to make it all the way to whatever actually allocates the Socket, then use `getLocalPort()` (directly or indirectly) on that. Generally this does require a bit of extra logic in the main code itself, e.g. see [equivalent in KafkaServer](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaServer.scala#L388-L393) to generate correct advertised port/url.

i also started cleaning this up elsewhere to fix up flaky tests, might be useful as a reference for how to get this working https://github.com/confluentinc/rest-utils/pull/90

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,321 @@
+        try {
+            producer.close();
+        } catch (Exception e) {
+            log.error("Could not shutdown producer ", e);
```
same as mentioned elsewhere, none of these errors should just be swallowed and logged since they won't be noticed, they should cause the test to fail (or have the test marked with expected failures if they are expected to fail)

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,321 @@
+            final long startMillis = System.currentTimeMillis();
+            long current = startMillis;
+            while (current - startMillis < maxDuration) {
+                ConsumerRecords<byte[], byte[]> rec = consumer.poll(Duration.ofMillis(maxDuration));
```
parameter to `poll` should be amount of time left, not `maxDuration`

wicknicks committed:
KAFKA-7503: Address reviewer comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Initialize RestServer before starting herder (#8)

This is necessary to keep the advertisedUrl consistent with the
host:port bound to by the SocketServer when a REST_PORT_CONFIG of 0 is
passed into RestServer from integration tests.

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Set REST_PORT_CONFIG to 0 for integration tests

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
MINOR: Suppress deprecation errors

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Replace premature exception with a warning

Signed-off-by: Arjun Satish <arjun@confluent.io>

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExampleConnectIntegrationTest.java:
```diff
@@ -0,0 +1,133 @@
+        connectorHandle.awaitRecords(CONSUME_MAX_DURATION_MS);
+
+        // at least one task should have initalized and executed
+        assertEquals("Incorrect task count in connector", 2, connect.getConnectorStatus(CONNECTOR_NAME).tasks().size());
```
You need something like `waitUntil` here, connector and task statuses depend on them actually being instantiated and the status being written to the status topic, and in multi-node clusters the other workers may need to read those updates before they are visible via the REST API.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/HerderProvider.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public void setHerder(Herder herder) {
+        initialized.countDown();
+        this.herder = herder;
```
the order here looks reversed from what it should, as is you risk reading off `this.herder` when still uninitialized

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ErrorHandlingIntegrationTest.java:
```diff
@@ -0,0 +1,224 @@
+
+        connectorHandle.taskHandle(TASK_ID).expectedRecords(EXPECTED_CORRECT_RECORDS);
+
+        connect.configureConnector("error-conn", props);
```
nit: redundant to class level constant for the connector name, repeated with `deleteConnector` as well

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/TaskHandle.java:
```diff
@@ -0,0 +1,126 @@
+            // this can happen if all tasks have not started, and more partitions are assigned
+            // to the first task. Subsequent rebalances should allocate the right number of
+            // partitions to each task.
+            log.warn("Expected only one partition. But, assigned " + numPartitions + ".");
```
If this happens, isn't `awaitRecords` potentially going to not work? If the first task is started, gets assigned multiple partitions, and consumes some records from partitions you aren't expecting to be assigned to it, presumably the test would fail?

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:
```diff
@@ -0,0 +1,330 @@
+    }
+
+    public void produce(String topic, Integer partition, String key, String value) {
+        producer.send(new ProducerRecord<>(topic, partition, key == null ? null : key.getBytes(), value == null ? null : value.getBytes()));
```
do you want to return the future, or possibly make these sync by default? might make it simpler to debug tests even if it is slower, and I think most tests probably wouldn't depend on using async sends.

ewencp commented:
@wicknicks This is looking very close, just a few last issues I found.

wicknicks committed:
KAFKA-7503: Add waitUntil and address reviewer comments

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp addressed your comments. could you take another pass? thanks so much!

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/HerderProvider.java:
```diff
@@ -0,0 +1,66 @@
+    public Herder get() {
+        try {
+            // wait up to 15 seconds
+            initialized.await(15, TimeUnit.SECONDS);
```
shouldn't we be checking the return value here? it seems like the case where we wait 15 seconds is not handled as described in the javadoc. probably not the end of the world since it'll just die with a null pointer exception somewhere later, but if handled here we could provide a better error message closer to the source of the issue.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectIntegrationTestUtils.java:
```diff
@@ -0,0 +1,51 @@
+    private static final Logger log = LoggerFactory.getLogger(ConnectIntegrationTestUtils.class);
+    private static final long CONNECTOR_SETUP_DURATION_MS = 100;
+
+    public static void waitUntil(Supplier<Boolean> condition, long waitForMillis, String errorMessage) {
```
There are some common test utilities in o.a.k.test.TestUtils which basically has this, with some additional variants for things like using a default timeout value. Your variant is a bit different in doing exponential backoff, though I think it's pretty rare we really expect to need that. Would be good to figure out if we really need something different here or we can just reuse the existing implementations. (in general, I'd prefer to avoid having different utilities for the same thing across all 3 major AK modules, though due to core being in scala I think we'll inevitably end up with a bit of duplication there until we can migrate to a java equivalent in core; but for Connect, I think we should only have stuff here that really is Connect-specific).

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExampleConnectIntegrationTest.java:
```diff
@@ -0,0 +1,137 @@
+        connect.start();
+
+        // get a handle to the connector
+        connectorHandle = RuntimeHandles.get().connectorHandle(CONNECTOR_NAME);
```
nit: its confusing that the order is different between the two integration tests and not necessarily obvious to an author if the order of the `RuntimeHandles` bits matter

wicknicks committed:
KAFKA-7503: Use waitForCondition from TestUtils; throw error if Latch doesn't count down

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7503: Fix call to RestServer#start in unit test

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@ewencp these are the changes made after your last comments: 
https://github.com/apache/kafka/pull/5516/commits/8b24c6fbf2806b6e3c6d764f212e392c8e204fff
https://github.com/apache/kafka/pull/5516/commits/027e68ee156e015bf3b38d14d529b5780009e221
the only changes different from your comments are:
1. making the herder field in `HerderProvider` volatile, for safe reads across threads.
2. fixes to a unit test after rebasing on `trunk`.
thanks!

ewencp reviewed (approved):
LGTM, merging to trunk and as far back as it'll cherry-pick cleanly and pass tests since this is just adding tests that would be useful on any branch. Thanks for all the iterations on this!

wicknicks commented:
Thanks for your time and patience on this, @ewencp! Much appreciated!
